[iOS] Date picker popover uses different background for arrow than the rest of view, toolbar sometimes clipped

lilyspiniolas opened the pull request:
#### 34f30da711476a783cd5bf41e806d5697ceb6220
<pre>
[iOS] Date picker popover uses different background for arrow than the rest of view, toolbar sometimes clipped
<a href="https://bugs.webkit.org/show_bug.cgi?id=290439">https://bugs.webkit.org/show_bug.cgi?id=290439</a>
<a href="https://rdar.apple.com/147361697">rdar://147361697</a>

Reviewed by Wenson Hsieh and Aditya Keerthi.

The UIVisualEffectView used by WKDatePickerPopoverView is not necessary
as the popover view is already using a blurred background by default.
This resulted in a double blur everywhere but the popover arrow, which
only had a single blur and thus caused a difference in appearance
compared to the rest of the view. The UIVisualEffectView has been removed,
and layout constraints have been updated accordingly. Additionally, the
toolbar&apos;s bottom margin is now adjusted when necessary in order to prevent
clipping.

* Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
(adjustedMarginIfNeeded):
(-[WKDatePickerPopoverView setupView:toolbarBottomMargin:]):
(-[WKDatePickerPopoverView bottomMarginForToolbar]):
(-[WKDatePickerPopoverView initWithDatePicker:]):
(-[WKDatePickerPopoverView estimatedMaximumPopoverSize]):

Canonical link: <a href="https://commits.webkit.org/292720@main">https://commits.webkit.org/292720@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/60d99c8678ac319dc8a81790bc171e72bb586ab3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96818 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16432 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6597 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101891 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47338 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16728 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24874 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73771 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30984 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99821 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12591 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54106 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12352 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46666 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82449 "Passed tests") | [⏳ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Tests-EWS "Waiting to run tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103914 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23886 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17422 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82819 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24137 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83633 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82208 "Found 3 new API test failures: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/close-after-print (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26869 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17368 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15627 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23848 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29003 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23507 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26987 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25248 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
```diff
@@ -113,10 +111,19 @@ - (void)setupView:(UIView *)pickerView toolbarBottomMargin:(CGFloat)toolbarBotto
         [[_accessoryView heightAnchor] constraintEqualToConstant:accessoryViewSize.height],
 #endif
-        [[_accessoryView bottomAnchor] constraintEqualToAnchor:[_backgroundView bottomAnchor]],
+        [[_accessoryView bottomAnchor]  constraintEqualToAnchor:self.bottomAnchor constant:-toolbarBottomMargin],
```
```suggestion
        [[_accessoryView bottomAnchor] constraintEqualToAnchor:self.bottomAnchor constant:-toolbarBottomMargin],
```

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
```diff
@@ -113,10 +111,19 @@ - (void)setupView:(UIView *)pickerView toolbarBottomMargin:(CGFloat)toolbarBotto
+- (CGFloat)bottomMarginForToolbar
+{
+    auto margin = toolbarBottomMarginSmall;
+    if (_datePicker && _datePicker.datePickerMode == UIDatePickerModeDateAndTime)
```
No need to `nil` check `_datePicker`.

pxlcoder replied:
(`_datePicker.datePickerMode` would evaluate to 0 in that case)

lilyspiniolas committed:
[iOS] Date picker popover uses different background for arrow than the rest of view, toolbar sometimes clipped
https://bugs.webkit.org/show_bug.cgi?id=290439
rdar://147361697

Reviewed by Wenson Hsieh and Aditya Keerthi.

The UIVisualEffectView used by WKDatePickerPopoverView is not necessary
as the popover view is already using a blurred background by default.
This resulted in a double blur everywhere but the popover arrow, which
only had a single blur and thus caused a difference in appearance
compared to the rest of the view. The UIVisualEffectView has been removed,
and layout constraints have been updated accordingly. Additionally, the
toolbar's bottom margin is now adjusted when necessary in order to prevent
clipping.

* Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
(adjustedMarginIfNeeded):
(-[WKDatePickerPopoverView setupView:toolbarBottomMargin:]):
(-[WKDatePickerPopoverView bottomMarginForToolbar]):
(-[WKDatePickerPopoverView initWithDatePicker:]):
(-[WKDatePickerPopoverView estimatedMaximumPopoverSize]):

Canonical link: https://commits.webkit.org/292720@main
